src/client/clouds.cpp

```diff
@@ -223,96 +231,197 @@ void Clouds::updateMesh()
+					v[3].Color = c_bottom;
+					v[0].Pos.set(-rx, ry, -rz);
+					v[1].Pos.set(rx, ry, -rz);
+					v[2].Pos.set(rx, 0, -rz);
+					v[3].Pos.set(-rx, 0, -rz);
+					break;
+				case 2: //right
+					if (INAREA(xi + 1, zi, m_cloud_radius_i)) {
+						u32 j = GETINDEX(xi + 1, zi, m_cloud_radius_i);
+						if (grid[j])
+							continue;
+					}
+					for (video::S3DVertex& vertex : v) {
+						vertex.Normal.set(1, 0, 0);
+					}
+					v[2].Color = c_bottom;
+					v[3].Color = c_bottom;
+					v[0].Pos.set(rx, ry, -rz);
+					v[1].Pos.set(rx, ry, rz);
+					v[2].Pos.set(rx, 0, rz);
+					v[3].Pos.set(rx, 0, -rz);
+					break;
+				case 3: // front
+					if (INAREA(xi, zi + 1, m_cloud_radius_i)) {
+						u32 j = GETINDEX(xi, zi + 1, m_cloud_radius_i);
+						if (grid[j])
+							continue;
+					}
+					for (video::S3DVertex& vertex : v) {
+						vertex.Normal.set(0, 0, -1);
+					}
+					v[2].Color = c_bottom;
+					v[3].Color = c_bottom;
+					v[0].Pos.set(rx, ry, rz);
+					v[1].Pos.set(-rx, ry, rz);
+					v[2].Pos.set(-rx, 0, rz);
+					v[3].Pos.set(rx, 0, rz);
+					break;
+				case 4: // left
+					if (INAREA(xi - 1, zi, m_cloud_radius_i)) {
+						u32 j = GETINDEX(xi - 1, zi, m_cloud_radius_i);
+						if (grid[j])
+							continue;
+					}
+					for (video::S3DVertex& vertex : v) {
+						vertex.Normal.set(-1, 0, 0);
+					}
+					v[2].Color = c_bottom;
+					v[3].Color = c_bottom;
+					v[0].Pos.set(-rx, ry, rz);
+					v[1].Pos.set(-rx, ry, -rz);
+					v[2].Pos.set(-rx, 0, -rz);
+					v[3].Pos.set(-rx, 0, rz);
+					break;
+				case 5: // bottom
+					for (video::S3DVertex& vertex : v) {
+						vertex.Color = c_bottom;
+						vertex.Normal.set(0, -1, 0);
+					}
+					v[0].Pos.set(rx, 0, rz);
+					v[1].Pos.set(-rx, 0, rz);
+					v[2].Pos.set(-rx, 0, -rz);
+					v[3].Pos.set(rx, 0, -rz);
+					break;
 				}
-				for (video::S3DVertex &vertex : v) {
-					vertex.Color = c_side_1;
-					vertex.Normal.set(0,0,-1);
-				}
-				v[0].Pos.set( rx, ry, rz);
-				v[1].Pos.set(-rx, ry, rz);
-				v[2].Pos.set(-rx,  0, rz);
-				v[3].Pos.set( rx,  0, rz);
-				break;
-			case 4: // left
-				if (INAREA(xi-1, zi, m_cloud_radius_i)) {
-					u32 j = GETINDEX(xi-1, zi, m_cloud_radius_i);
-					if(grid[j])
-						continue;
-				}
-				for (video::S3DVertex &vertex : v) {
-					vertex.Color = c_side_2;
-					vertex.Normal.set(-1,0,0);
-				}
-				v[0].Pos.set(-rx, ry, rz);
-				v[1].Pos.set(-rx, ry,-rz);
-				v[2].Pos.set(-rx,  0,-rz);
-				v[3].Pos.set(-rx,  0, rz);
-				break;
-			case 5: // bottom
-				for (video::S3DVertex &vertex : v) {
-					vertex.Color = c_bottom;
-					vertex.Normal.set(0,-1,0);
+
+				v3f pos(p0.X, m_params.height * BS, p0.Y);
+
+				for (video::S3DVertex& vertex : v) {
+					vertex.Pos += pos;
+					mb->Vertices.push_back(vertex);
 				}
-				v[0].Pos.set( rx,  0, rz);
-				v[1].Pos.set(-rx,  0, rz);
-				v[2].Pos.set(-rx,  0,-rz);
-				v[3].Pos.set( rx,  0,-rz);
-				break;
 			}
+		}
+		// Standard clouds
```

Comment:
I hope this can be de-duplicated in some way because it's largely identical
e.g. by moving the if inside the loop. This function doesn't run every frame and shouldn't be totally performance-critical.